controller/src/main/java/org/jboss/as/controller/ValidateModelStepHandler.java

```diff
@@ -38,33 +44,48 @@ private ValidateModelStepHandler() {
+        final Resource resource = loadResource(context);
+        if (resource == null) {
+            return;
+        }
+
         final ModelNode model = resource.getModel();
         final ImmutableManagementResourceRegistration resourceRegistration = context.getResourceRegistration();
         final Set<String> attributeNames = resourceRegistration.getAttributeNames(PathAddress.EMPTY_ADDRESS);
-        for (String attributeName : attributeNames) {
+        for (final String attributeName : attributeNames) {
             final boolean has = model.hasDefined(attributeName);
-            AttributeAccess access = context.getResourceRegistration().getAttributeAccess(PathAddress.EMPTY_ADDRESS, attributeName);
+            final AttributeAccess access = context.getResourceRegistration().getAttributeAccess(PathAddress.EMPTY_ADDRESS, attributeName);
             if (access.getStorageType() != AttributeAccess.Storage.CONFIGURATION){
                 continue;
             }
-            AttributeDefinition attr = access.getAttributeDefinition();
+            final AttributeDefinition attr = access.getAttributeDefinition();
             if (!has && isRequired(attr, model)) {
-                throw new OperationFailedException(ControllerLogger.ROOT_LOGGER.required(attributeName));
+                attemptReadMissingAttributeValueFromHandler(context, access, attributeName, new ErrorHandler() {
+                    @Override
+                    public void throwError() throws OperationFailedException {
+                        throw new OperationFailedException(ControllerLogger.ROOT_LOGGER.required(attributeName));
+                    }});
             }
             if (!has) {
                 continue;
             }
 
             if (attr.getRequires() != null) {
-                for (String required : attr.getRequires()) {
+                for (final String required : attr.getRequires()) {
                     if (!model.hasDefined(required)) {
-                        throw ControllerLogger.ROOT_LOGGER.requiredAttributeNotSet(required, attr.getName());
+                        attemptReadMissingAttributeValueFromHandler(context, access, attributeName, new ErrorHandler() {
+                            @Override
+                            public void throwError() throws OperationFailedException {
+                                throw ControllerLogger.ROOT_LOGGER.requiredAttributeNotSet(required, attr.getName());
+                            }});
                     }
                 }
             }
 
             if (!isAllowed(attr, model)) {
+                //TODO should really use attemptReadMissingAttributeValueFromHandler() to make this totally good, but the
+                //overhead might be bigger than is worth at the moment since we would have to invoke the extra steps for
+                //every single attribute not found (and not found should be the normal).
```

Comment:
I agree with the thinking in the comment.

This isn't an issue with this PR, but it seems that isAllowed and this block guarded by the call to isAllowed are doing redundant work. There could be a single loop to create the StringBuilder, and fail if sb != null.